modules/core/src/main/java/org/apache/ignite/internal/managers/encryption/CacheGroupPageScanner.java

```diff
@@ -0,0 +1,453 @@
+                    cpWaitGrps.addAll(completeCandidates);
+
+                    return;
+                }
+
+                lock.lock();
+
+                try {
+                    for (GroupScanTask grpScanTask : completeCandidates) {
+                        grps.remove(grpScanTask.groupId());
+
+                        grpScanTask.onDone();
+
+                        if (log.isInfoEnabled())
+                            log.info("Cache group reencryption is finished [grpId=" + grpScanTask.groupId() + "]");
+                    }
+
+                    if (!grps.isEmpty())
+                        return;
+
+                    ((GridCacheDatabaseSharedManager)ctx.cache().context().database()).
+                        removeCheckpointListener(this);
+                }
+                finally {
+                    lock.unlock();
+                }
+            }
+        );
+    }
+
+    /** {@inheritDoc} */
+    @Override public void beforeCheckpointBegin(Context cpCtx) {
+        // No-op.
+    }
+
+    /** {@inheritDoc} */
+    @Override public void onMarkCheckpointBegin(Context ctx) {
+        // No-op.
+    }
+
+    /**
+     * Schedule scanning partitions.
+     *
+     * @param grpId Cache group ID.
+     */
+    public IgniteInternalFuture<Void> schedule(int grpId) throws IgniteCheckedException {
+        CacheGroupContext grp = ctx.cache().cacheGroup(grpId);
+
+        if (grp == null) {
+            if (log.isDebugEnabled())
+                log.debug("Skip reencryption, cache group was destroyed [grp=" + grpId + "]");
+
+            return new GridFinishedFuture<>();
+        }
+
+        lock.lock();
+
+        try {
+            if (stopped)
+                throw new NodeStoppingException("Operation has been cancelled (node is stopping).");
+
+            if (grps.isEmpty())
+                ((GridCacheDatabaseSharedManager)ctx.cache().context().database()).addCheckpointListener(this);
+
+            GroupScanTask prevState = grps.get(grpId);
+
+            if (prevState != null && !prevState.isDone()) {
+                if (log.isDebugEnabled())
+                    log.debug("Reencryption already scheduled [grpId=" + grpId + "]");
+
+                return prevState;
+            }
+
+            Set<Integer> parts = new HashSet<>();
+
+            forEachPageStore(grp, new IgniteInClosureX<Integer>() {
+                @Override public void applyx(Integer partId) {
+                    if (ctx.encryption().getEncryptionState(grpId, partId) == 0) {
+                        if (log.isDebugEnabled())
+                            log.debug("Skipping partition reencryption [grp=" + grpId + ", p=" + partId + "]");
+
+                        return;
+                    }
+
+                    parts.add(partId);
+                }
+            });
+
+            GroupScanTask grpScan = new GroupScanTask(grp, parts);
+
+            ctx.getSystemExecutorService().submit(grpScan);
```

Comment:
Here task for each cache group can be executed by different threads. Let's use single-threaded execution here.

Reply:
Added single threaded executor